File: atoi.py
#!/usr/bin/env python

__author__ = 'Rio'

def atoi(string):
    ''' atoi
    '''
    if not string: return string

    INT_MAX, INT_MIN = 2147483647, -2147483648

    # trim all whitespaces
    string = string.strip()

    flag = False
    index = 0
    if string[index] == '-':
        flag = True
        index += 1
    elif string[index] == '+':
        index += 1

    result = 0
    while index < len(string) and string[index] >= '0' and string[index] <= '9':
        result = result * 10 + int(string[index])
        index += 1

    if flag:
        result = -result

    if result > INT_MAX:
        return INT_MAX
    elif result < INT_MIN:
        return INT_MIN


    return result

File: java/Atoi.java
public class Atoi {
    public int atoi(String str) {
        if (str == null) {
            return 0;
        }

        str = str.trim();
        if (str.length() == 0) {
            return 0;
        }

        int index = 0;
        boolean negative = false;
        if (str.charAt(index) == '-') {
            negative = true;
            index++;
        } else if (str.charAt(index) == '+') {
            index++;
        }

        long result = 0;
        while (index < str.length() && str.charAt(index) >= '0' && str.charAt(index) <= '9') {
            result = result * 10 + (str.charAt(index) - '0');
            index++;
        }

        if (negative) result = -result;
        if (result >= Integer.MAX_VALUE) {
            return Integer.MAX_VALUE;
        } else if (result <= Integer.MIN_VALUE) {
            return Integer.MIN_VALUE;
        }

        return (int)result;
    }
}
